### PR TITLE
Karl + Nginx.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Karl/karl.h [hunk 1]
```diff
@@ -192,8 +195,90 @@ CURRENT_STRUCT_T(KarlPersistedKeepalive) {
...
+        route_prefix(route_prefix),
+        validate_full_config_on_every_update(validate_full_config_on_every_update) {}
+};
+
+template <class STORAGE>
+class KarlNginxManager {
+ protected:
+  explicit KarlNginxManager(STORAGE& storage, const KarlNginxParameters& nginx_parameters)
+      : storage_(storage),
+        using_nginx_(!nginx_parameters.config_file.empty()),
```
Comment (dkorolev) [inline Karl/karl.h]: `has_nginx_config_file_` ?

File: Karl/karl.h [hunk 2]
```diff
@@ -192,8 +195,90 @@ CURRENT_STRUCT_T(KarlPersistedKeepalive) {
...
+            need_to_update_nginx_ = false;
+          } else {
+#ifdef EXTRA_KARL_LOGGING
+            std::cerr << "Nginx config reload failed." << std::endl;
+#endif
+          }
+        } catch (const current::FileException& e) {
+#ifdef EXTRA_KARL_LOGGING
+          std::cerr << "Cannot write Nginx config file to '" << nginx_parameters_.config_file
+                    << "': " << e.what() << std::endl;
```
Comment (dkorolev) [inline Karl/karl.h]: `s/what()/What()/g` for Current exceptions.

File: Karl/karl.h [hunk 3]
```diff
@@ -192,8 +195,90 @@ CURRENT_STRUCT_T(KarlPersistedKeepalive) {
...
+  virtual ~KarlNginxManager() {
+    if (using_nginx_) {
+      FileSystem::RmFile(nginx_parameters_.config_file, FileSystem::RmFileParameters::Silent);
+      nginx::NginxInvoker().ReloadConfig();
+    }
+  }
+
+ protected:
+  STORAGE& storage_;
+  bool using_nginx_;
```
Comment (dkorolev) [inline Karl/karl.h]: `const` ?

File: Karl/karl.h [hunk 4]
```diff
@@ -192,8 +195,90 @@ CURRENT_STRUCT_T(KarlPersistedKeepalive) {
...
+  KarlNginxParameters nginx_parameters_;
```
Comment (dkorolev) [inline Karl/karl.h]: `const` ?

File: Karl/karl.h [hunk 5]
```diff
@@ -227,14 +316,68 @@ class GenericKarl final {
...
+    return services_keepalive_time_cache_.size();
   }
 
+  storage_t& InternalExposeStorage() { return storage_; }
+
  private:
+  void StateUpdateThread() {
+    while (!destructing_) {
+      const auto now = current::time::Now();
+      std::set<std::string> timeouted_codenames;
```
Comment (dkorolev) [inline Karl/karl.h]: `unordered_set` ?

File: Karl/karl.h [hunk 6]
```diff
@@ -578,10 +747,16 @@ class GenericKarl final {
                  std::move(r)).Detach();
   }
 
+  std::atomic_bool destructing_;
+  std::unordered_map<std::string, std::chrono::microseconds> services_keepalive_time_cache_;
+  mutable std::mutex services_keepalive_cache_mutex_;
+  std::condition_variable active_services_condition_variable_;
+
   const std::string external_url_;
   const std::chrono::microseconds up_threshold_;
```
Comment (dkorolev) [inline Karl/karl.h]: `s/up_threshold_/service_timeout_interval_` ?

File: Karl/karl.h [hunk 7]
```diff
@@ -227,14 +316,68 @@ class GenericKarl final {
...
+      }
+      if (!timeouted_codenames.empty()) {
+        storage_.ReadWriteTransaction([&timeouted_codenames](MutableFields<storage_t> fields) -> void {
+          for (const auto& codename : timeouted_codenames) {
+            const auto& current_claire_info = fields.claires[codename];
+            ClaireInfo claire;
+            if (Exists(current_claire_info)) {
+              claire = Value(current_claire_info);
+            } else {
+              claire.codename = codename;
```
Comment (dkorolev) [inline Karl/karl.h]: :+1: perhaps mark this Claire as a zombie, even by the simplest way possible, by writing some `"zombie"` as its IP address?

File: Karl/karl.h [hunk 8]
```diff
@@ -313,17 +463,17 @@ class GenericKarl final {
             }
           }();
 
+          const auto now = current::time::Now();
```
Comment (dkorolev) [inline Karl/karl.h]: Any reason why these three lines moved above the next block? :-)

File: Karl/karl.h [hunk 9]
```diff
@@ -405,7 +560,17 @@ class GenericKarl final {
                          }
                          return Response("OK\n");
                        },
-                       std::move(r)).Detach();
+                       std::move(r)).Wait();
+          {
+            std::lock_guard<std::mutex> lock(services_keepalive_cache_mutex_);
+            if (services_keepalive_time_cache_.count(codename) == 0u) {
```
Comment (dkorolev) [inline Karl/karl.h]: As you initialize it below, unless `0u` is a valid value, might as well do `auto& placeholder = services_keepalive_time_cache_[codename];` and compare it to zero.

File: Karl/karl.h [hunk 10]
```diff
@@ -391,8 +544,10 @@ class GenericKarl final {
                            }
                            claire.codename = codename;
 
-                           // TODO(mzhurovich): This one should work via `nginx`, I'd assume.
-                           claire.url_status_page_proxied = external_url_ + "proxied/" + codename;
+                           if (using_nginx_) {
+                             claire.url_status_page_proxied =
```
Comment (dkorolev) [inline Karl/karl.h]: Suggest we remove it nah from the persisted storage.

File: Karl/karl.h [hunk 11]
```diff
@@ -546,7 +711,11 @@ class GenericKarl final {
                            blob.unresolved_dependencies.push_back(dep.StatusPageURL());
                          }
                        }
-                       blob.url_status_page_proxied = external_url_ + "proxied/" + codename;
+                       // TODO(dk+mz): This should not be made twice, I assume.
```
Comment (dkorolev) [inline Karl/karl.h]: ^ Suggest we remove `url_status_page_proxied` nah from the persisted storage.

File: Karl/test.cc [hunk 12]
```diff
@@ -51,6 +51,12 @@ DEFINE_uint16(karl_generator_test_port, PickPortForUnitTest(), "Local test port
 DEFINE_uint16(karl_is_prime_test_port, PickPortForUnitTest(), "Local test port for the `is_prime` service.");
 DEFINE_uint16(karl_annotator_test_port, PickPortForUnitTest(), "Local test port for the `annotator` service.");
 DEFINE_uint16(karl_filter_test_port, PickPortForUnitTest(), "Local test port for the `filter` service.");
+DEFINE_uint16(karl_nginx_port,
+              PickPortForUnitTest(),
+              "Local port for Nginx to listen for proxying requests to Claires.");
+DEFINE_string(karl_nginx_config_file,
```
Comment (dkorolev) [inline Karl/test.cc]: An example snippet of how to include this local nginx config would be nice.

File: Karl/test.cc [hunk 13]
```diff
@@ -312,6 +318,144 @@ TEST(Karl, Deregister) {
...
+              HTTP(GET(Printf("http://localhost:%d?from=0&full&active_only", FLAGS_karl_test_port))).body));
+      EXPECT_EQ(1u, status.size());
+      ASSERT_TRUE(status.count("127.0.0.1")) << JSON(status);
+      auto& per_ip_services = status["127.0.0.1"].services;
+      EXPECT_EQ(1u, per_ip_services.size());
+      EXPECT_EQ("generator", per_ip_services[generator.ClaireCodename()].service);
+    }
+  }
+
+  // All services should be deregistered at this moment.
```
Comment (dkorolev) [inline Karl/test.cc]: Nit: at -> by.

File: Karl/test.cc [hunk 14]
```diff
@@ -312,6 +318,144 @@ TEST(Karl, Deregister) {
...
+            HTTP(GET(Printf("http://localhost:%d?from=0&full&active_only", FLAGS_karl_test_port))).body));
+    EXPECT_EQ(0u, status.size()) << JSON(status);
+  }
+}
+
+TEST(Karl, DeregisterWithNginx) {
+  current::time::ResetToZero();
+
+  // Run the test only if `karl_nginx_config_file` flag is set.
+  if (FLAGS_karl_nginx_config_file.empty()) {
```
Comment (dkorolev) [inline Karl/test.cc]: Nit: Perform this check before `ResetToZero()` ?

File: Karl/test.cc [hunk 15]
```diff
@@ -312,6 +318,144 @@ TEST(Karl, Deregister) {
...
+                             nginx_parameters);
+  const current::karl::Locator karl_locator(Printf("http://localhost:%d", FLAGS_karl_test_port));
+
+  // No services registered.
+  {
+    unittest_karl_status_t status;
+    ASSERT_NO_THROW(
+        status = ParseJSON<unittest_karl_status_t>(
+            HTTP(GET(Printf("http://localhost:%d?from=0&full&active_only", FLAGS_karl_test_port))).body));
+    EXPECT_EQ(0u, status.size());
```
Comment (dkorolev) [inline Karl/test.cc]: `EXPECT_TRUE(status.empty()) << JSON(status);`

File: Karl/test.cc [hunk 16]
```diff
@@ -312,6 +318,144 @@ TEST(Karl, Deregister) {
...
+  {
+    const karl_unittest::ServiceGenerator generator(
+        FLAGS_karl_generator_test_port, std::chrono::microseconds(1000), karl_locator);
+    // The `generator` service is registered.
+    {
+      unittest_karl_status_t status;
+      ASSERT_NO_THROW(
+          status = ParseJSON<unittest_karl_status_t>(
+              HTTP(GET(Printf("http://localhost:%d?from=0&full&active_only", FLAGS_karl_test_port))).body));
+      EXPECT_EQ(1u, status.size());
```
Comment (dkorolev) [inline Karl/test.cc]: ` << JSON(status);`

File: Karl/test.cc [hunk 17]
```diff
@@ -312,6 +318,144 @@ TEST(Karl, Deregister) {
...
+      EXPECT_EQ("generator", per_ip_services[generator.ClaireCodename()].service);
+      ASSERT_TRUE(Exists(per_ip_services[generator.ClaireCodename()].url_status_page_proxied));
+      generator_proxied_status_url = Value(per_ip_services[generator.ClaireCodename()].url_status_page_proxied);
+      EXPECT_EQ(karl_nginx_base_url + "/proxied/" + generator.ClaireCodename(), generator_proxied_status_url);
+    }
+
+    // Check that `generator`'s status page is accessible via Nginx.
+    {
+      current::karl::ClaireStatus status;
+      while (true) {
```
Comment (dkorolev) [inline Karl/test.cc]: Need a comment here, along the lines of `must wait for nginx config reload to take effect.`.

File: Karl/test.cc [hunk 18]
```diff
@@ -312,6 +318,144 @@ TEST(Karl, Deregister) {
...
+          }
+        } catch (const current::net::NetworkException& e) {
+        }
+      }
+      EXPECT_EQ("generator", status.service);
+    }
+
+    {
+      const karl_unittest::ServiceIsPrime is_prime(FLAGS_karl_is_prime_test_port, karl_locator);
+      // `generator` and `is_prime` services are registered.
```
Comment (dkorolev) [inline Karl/test.cc]: The :-)

File: Karl/karl.h (hunk 2 above)
Comment (mzhurovich) [inline Karl/karl.h]: :+1:

File: Karl/karl.h (hunk 1 above)
Comment (mzhurovich) [inline Karl/karl.h]: :+1:

File: Karl/karl.h (hunk 3 above)
Comment (mzhurovich) [inline Karl/karl.h]: :+1:

File: Karl/karl.h (hunk 4 above)
Comment (mzhurovich) [inline Karl/karl.h]: :+1:

File: Karl/test.cc [hunk 19]
```diff
@@ -322,6 +466,165 @@ TEST(Karl, Deregister) {
...
+                       "/",
+                       karl_nginx_base_url,
+                       nginx_parameters);
+  const current::karl::Locator karl_locator(Printf("http://localhost:%d", FLAGS_karl_test_port));
+
+  current::karl::ClaireStatus claire;
+  claire.service = "unittest";
+  claire.codename = "ABCDEF";
+  claire.local_port = PickPortForUnitTest();
+  // Register fake service.
```
Comment (dkorolev) [inline Karl/test.cc]: Nit: a fake service.

File: Karl/karl.h (hunk 5 above)
Comment (mzhurovich) [inline Karl/karl.h]: :+1:

File: Karl/karl.h (hunk 6 above)
Comment (mzhurovich) [inline Karl/karl.h]: :+1:

File: Karl/test.cc [hunk 20]
```diff
@@ -322,6 +466,165 @@ TEST(Karl, Deregister) {
...
+  HTTP(claire.local_port).Register("/.current", [](Request r) { r("GOTIT\n"); });
+
+  std::string proxied_url;
+  {
+    const std::string keepalive_url = Printf("%s?codename=%s&port=%d",
```
Comment (dkorolev) [inline Karl/test.cc]: Why not declare and initialize `current::karl::ClaireStatus claire;` in a very narrow scope running the POST on the next line, and use a bare `"ABCDEF`" string everywhere else? No need to DRY the test.

File: Karl/test.cc [hunk 21]
```diff
@@ -322,6 +466,165 @@ TEST(Karl, Deregister) {
...
+                            .ReadOnlyTransaction([&](ImmutableFields<unittest_karl_t::storage_t> fields) {
+                              ASSERT_TRUE(Exists(fields.claires[claire.codename]));
+                              EXPECT_EQ(current::karl::ClaireRegisteredState::DisconnectedByTimeout,
+                                        Value(fields.claires[claire.codename]).registered_state);
+                            })
+                            .Go();
+    EXPECT_TRUE(WasCommitted(result));
+  }
+
+  // This is a bit flaky.
```
Comment (dkorolev) [inline Karl/test.cc]: Why is it? As nginx reloads the config, this route should return 404; why won't it?

File: Karl/karl.h (hunk 8 above)
Comment (mzhurovich) [inline Karl/karl.h]: More accurate time skew calculation :)
(I also used it in some intermediate version to update status along with the `Publish` to the stream, but then changed the logic.)

File: Karl/test_service/generator.h [hunk 22]
```diff
@@ -80,10 +80,12 @@ class ServiceGenerator final {
   void Thread() {
     while (!destructing_) {
 #ifdef CURRENT_MOCK_TIME
-      current::time::SetNow(std::chrono::microseconds((current_value_ + 1) * 1000ull * 1000ull),
-                            std::chrono::microseconds((current_value_ + 2) * 1000ull * 1000ull - 1));
-#endif
+      // Removed `SetNow` to avoid problems with service timeout tests -- M.Z.
```
Comment (dkorolev) [inline Karl/test_service/generator.h]: This comment line is safe to remove.
:+1:

**Comment by dkorolev** [comment]
LGTM++!

File: Karl/karl.h (hunk 8 above)
Comment (dkorolev) [inline Karl/karl.h]: Perhaps call it some `now_for_time_skew_computation` then? A bit confusing as is, with two `now` variables declared one in the outer and one in the inner scope.

File: Karl/karl.h (hunk 10 above)
Comment (mzhurovich) [inline Karl/karl.h]: :+1:

File: Karl/karl.h (hunk 9 above)
Comment (mzhurovich) [inline Karl/karl.h]: :ok_hand:

File: Karl/test.cc (hunk 16 above)
Comment (mzhurovich) [inline Karl/test.cc]: :+1:

File: Karl/test.cc (hunk 13 above)
Comment (mzhurovich) [inline Karl/test.cc]: :+1:

File: Karl/test.cc (hunk 15 above)
Comment (mzhurovich) [inline Karl/test.cc]: :+1:

File: Karl/test.cc (hunk 17 above)
Comment (mzhurovich) [inline Karl/test.cc]: :+1:

File: Karl/test.cc (hunk 18 above)
Comment (mzhurovich) [inline Karl/test.cc]: :+1:

File: Karl/test.cc (hunk 19 above)
Comment (mzhurovich) [inline Karl/test.cc]: :+1:

**Pull request merged**